## Capstone project - Week 3 - Toronto Neighborhood

In [4]:
import requests
import pandas as pd
import numpy as np

In [5]:
# since new url has changed the format of the postal code table in Wiki page, we are using the old url
url = "https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969"
old_url = requests.get(url)
old_url

<Response [200]>

In [6]:
# we will scrape the data from html to text format using pandas
raw_data = pd.read_html(old_url.text)
raw_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

In [7]:
# now we will make the text data into a pandas data frame keeping only the postal codes section
raw_data = raw_data[0]
raw_data.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [8]:
# we will remove the not assigned rows and reset the index
df = raw_data[raw_data["Borough"] != "Not assigned"]
df = df.reset_index()
df.head()

index Postal Code           Borough  \
0      2         M3A        North York   
1      3         M4A        North York   
2      4         M5A  Downtown Toronto   
3      5         M6A        North York   
4      6         M7A  Downtown Toronto   

                                 Neighbourhood  
0                                    Parkwoods  
1                             Victoria Village  
2                    Regent Park, Harbourfront  
3             Lawrence Manor, Lawrence Heights  
4  Queen's Park, Ontario Provincial Government

In [9]:
# removing the old index
df.drop(['index'], axis = 'columns', inplace = True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [26]:
df.shape

(103, 3)

**now we have 103 postal codes of all the borough and neighborhoods of Toronto**

In [10]:
# now we will download latitude & longitude of all the postal codes, using the link below
data = pd.read_csv("https://cocl.us/Geospatial_data")
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [28]:
data.shape

(103, 3)

In [11]:
# since both the data set has 103 postal codes, we will assume they are same, and join the datasets using the postal codes
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [12]:
combined_data.shape

(103, 5)

**We have a complete dataset of all 103 postal codes. Next we will map the neighborhood on the map of Toronto, using folium pacakge**

In [31]:
!pip install geopy
from geopy.geocoders import Nominatim

In [32]:
# getting the lat & lon of Toronto using geocoder
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [33]:
!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.7-main

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |           1_llvm           5 KB  conda-forge
    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge
    _pytorch_select-0.2        |            gpu_0           2 KB
    absl-py-0.12.0             |     pyhd8ed1ab_0          96 KB  conda-forge
    aiohttp-3.7.4              |   py37h5e8e339_0         632 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
   

In [34]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

**Now we will initialize Foursqare API and 'explore' the venue data for all the Neighborhoods**

In [13]:
CLIENT_ID = 'X44MXRRXXIV00CMCETPJWZYUGN435BVLP4MECN0VUDBFMQRG' 
CLIENT_SECRET = 'HVFUVOSULDT4MWDJOXBSLXTO1GFC3F4ZWKHRAR41Q0RGSY2T'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: X44MXRRXXIV00CMCETPJWZYUGN435BVLP4MECN0VUDBFMQRG
CLIENT_SECRET:HVFUVOSULDT4MWDJOXBSLXTO1GFC3F4ZWKHRAR41Q0RGSY2T


In [14]:
# we are creating a function to get all the venue catagories for all the neighborhoods
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [15]:
# collecting the venues in Toronto for all neighborhoods, within 500 mtrs radius
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [16]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [17]:

venues_in_toronto.shape

(1326, 5)

So we have total 1326 venues in all the neighborhoods of Toronto

In [18]:
# checking by venue catagories with max frequency
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                                     
Accessories Store                               Lawrence Manor, Lawrence Heights   
Adult Boutique                                              Church and Wellesley   
Airport                                                                Downsview   
Airport Food Court             CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge                 CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service                CN Tower, King and Spadina, Railway Lands, Har...   
Airport Terminal               CN Tower, King and Spadina, Railway Lands, Har...   
American Restaurant                     Toronto Dominion Centre, Design Exchange   
Antique Shop                                       High Park, The Junction South   
Aquarium                       Harbourfront East, Union Station, Toronto Islands   
Art Gallery                             Toronto Dominion Centre, Design Exchange   
Art Museum                                                    Central Bay Street   
Arts & Crafts Store                                  Willowdale, Willowdale East   
Asian Restaurant                                        Richmond, Adelaide, King   
Athletics & Sports                                              Woodbine Heights   
Auto Garage                                                    Wexford, Maryvale   
BBQ Joint                                    The Annex, North Midtown, Yorkville   
Baby Store                                                              Christie   
Bagel Shop                     Summerhill West, Rathnelly, South Hill, Forest...   
Bakery                                                         Wexford, Maryvale   
Bank                                                 Willowdale, Willowdale East   
Bar                                               University of Toronto, Harbord   
Baseball Field                 Old Mill South, King's Mill Park, Sunnylea, Hu...   
Basketball Court                        North Park, Maple Leaf Park, Upwood Park   
Basketball Stadium             Harbourfront East, Union Station, Toronto Islands   
Beer Bar                                          University of Toronto, Harbord   
Beer Store                                                      Woodbine Heights   
Belgian Restaurant                     Kensington Market, Chinatown, Grange Park   
Bike Shop                                                                Leaside   
Bistro                         Harbourfront East, Union Station, Toronto Islands   
...                                                                          ...   
Sporting Goods Shop                            North Toronto West, Lawrence Park   
Sports Bar                                                               Leaside   
Stadium                             Brockton, Parkdale Village, Exhibition Place   
Stationery Store                                                 Studio District   
Steakhouse                                           Willowdale, Willowdale East   
Supermarket                                          Willowdale, Willowdale West   
Supplement Shop                Mimico NW, The Queensway West, South of Bloor,...   
Sushi Restaurant                                     Willowdale, Willowdale East   
Swim School                                                        Lawrence Park   
Tailor Shop                                                       Stn A PO Boxes   
Taiwanese Restaurant                                 St. James Town, Cabbagetown   
Tanning Salon                  Mimico NW, The Queensway West, South of Bloor,...   
Tea Room                               Kensington Market, Chinatown, Grange Park   
Tennis Court                                                  Runnymede, Swansea   
Thai Restaurant                                                  Studio District   
Theater                            

So there are 235 types of venues, Accessories Store being at Top and Yoga Studio at the bottom

In [19]:
# now we will get dummies for all the venue catagories

toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
5                     0               0        0                   0   
6                     0               0        0                   0   
7                     0               0        0                   0   
8                     0               0        0                   0   
9                     0               0        0                   0   
10                    0               0        0                   0   
11                    0               0        0                   0   
12                    0               0        0                   0   
13                    0               0        0                   0   
14                    0               0        0                   0   
15                    0               0        0                   0   
16                    0               0        0                   0   
17                    0               0        0                   0   
18                    0               0        0                   0   
19                    0               0        0                   0   
20                    0               0        0                   0   
21                    0               0        0                   0   
22                    0               0        0                   0   
23                    0               0        0                   0   
24                    0               0        0                   0   
25                    0               0        0                   0   
26                    0               0        0                   0   
27                    0               0        0                   0   
28                    0               0        0                   0   
29                    0               0        0                   0   
...                 ...             ...      ...                 ...   
1296                  0               0        0                   0   
1297                  0               0        0                   0   
1298                  0               0        0                   0   
1299                  0               0        0                   0   
1300                  0               0        0                   0   
1301                  0               0        0                   0   
1302                  0               0        0                   0   
1303                  0               0        0                   0   
1304                  0               0        0                   0   
1305                  0               0        0                   0   
1306                  0               0        0                   0   
1307                  0               0        0                   0   
1308                  0               0        0                   0   
1309                  0               0        0                   0   
1310                  0               0        0                   0   
1311                  0               0        0                   0   
1312                  0               0        0                   0   
1313                  0               0        0                   0   
1314                  0               0        0                   0   
1315                  0               0        0                   0   
1316                  0               0        0                   0   
1317                  0               0        0                   0   
1318                  0               0        0                   0   
1319                  0               0       

Here we have generated dummies for all the 1326 venues in Toronto Neighborhood

In [20]:
# Adding the Neighborhood column in the encoded dataset

toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  \
0                    0             0  ...      0              0   
1                    0             0  ...      0              0   
2                    0             0  ...      0              0   
3                    0             0  ...      0              0   
4                    0             0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 236 columns]

We will now group the data by Neighborhoods and get the mean of venue catagories for each neighborhood

In [21]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0             0.0      0.0                 0.0             0.0   
1             0.0      0.0                 0.0             0.0   
2             0.0      0.0                 0.0             0.0   
3             0.0      0.0                 0.0             0.0   
4             0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0              0.0               0.0             0.000000           0.0  ...   
1              0.0               0.0             0.000000           0.0  ...   
2              0.0               0.0             0.000000           0.0  ...   
3              0.0               0.0             0.000000           0.0  ...   
4              0.0               0.0             0.043478           0.0  ...   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0    0.0            0.0                            0.0               0.0   
1    0.0            0.0                            0.0               0.0   
2    0.0            0.0                            0.0               0.0   
3    0.0            0.0                            0.0               0.0   
4    0.0            0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 236 columns]

Now we will get the top most venues for each neighborhood

In [22]:
# first we will make a funstion to get the top most venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [23]:
# now plug in the function to get top 5 venues for each neighborhood
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood  \
0                                           Agincourt   
1                              Alderwood, Long Branch   
2     Bathurst Manor, Wilson Heights, Downsview North   
3                                     Bayview Village   
4                   Bedford Park, Lawrence Manor East   
5                                         Berczy Park   
6                         Birch Cliff, Cliffside West   
7        Brockton, Parkdale Village, Exhibition Place   
8   Business reply mail Processing Centre, South C...   
9   CN Tower, King and Spadina, Railway Lands, Har...   
10                                Caledonia-Fairbanks   
11              Canada Post Gateway Processing Centre   
12                                          Cedarbrae   
13                                 Central Bay Street   
14                                           Christie   
15                               Church and Wellesley   
16            Clarks Corners, Tam O'Shanter, Sullivan   
17    Cliffside, Cliffcrest, Scarborough Village West   
18                     Commerce Court, Victoria Hotel   
19                                         Davisville   
20                                   Davisville North   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn   
22                                          Don Mills   
23  Dorset Park, Wexford Heights, Scarborough Town...   
24                                          Downsview   
25                       Dufferin, Dovercourt Village   
26      East Toronto, Broadview North (Old East York)   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
28                       Fairview, Henry Farm, Oriole   
29             First Canadian Place, Underground city   
..                                                ...   
66                                           Roselawn   
67             Rouge Hill, Port Union, Highland Creek   
68                                 Runnymede, Swansea   
69  Runnymede, The Junction, Weston-Pellam Park, C...   
70                                Scarborough Village   
71  South Steeles, Silverstone, Humbergate, Jamest...   
72                                     St. James Town   
73                        St. James Town, Cabbagetown   
74                      Steeles West, L'Amoreaux West   
75                                     Stn A PO Boxes   
76                                    Studio District   
77  Summerhill West, Rathnelly, South Hill, Forest...   
78                The Annex, North Midtown, Yorkville   
79                                        The Beaches   
80                       The Danforth West, Riverdale   
81      The Kingsway, Montgomery Road, Old Mill North   
82                                   Thorncliffe Park   
83           Toronto Dominion Centre, Design Exchange   
84                     University of Toronto, Harbord   
85                                   Victoria Village   
86  West Deane Park, Princess Gardens, Martin Grov...   
87                                          Westmount   
88                                             Weston   
89                                  Wexford, Maryvale   
90                            Willowdale, Newtonbrook   
91                        Willowdale, Willowdale East   
92                        Willowdale, Willowdale West   
93                                             Woburn   
94                                   Woodbine Heights   
95                                    York Mills West   

        1st Most Common Venue  2nd Most Common Venue    3rd Most Common Venue  \
0   Latin American Restaurant                 Lounge             Skating Rink   
1                 Pizza Place            Coffee Shop                      Pub   
2                 Coffee Shop                   Bank            Shopping Mall   
3                        Café     Chinese Restaurant      Japanese Restaurant   
4              Sandwich Place     Italian Restaurant              Coffee Shop   
5                 C

**Observation** there are only 96 neighborhoods, this is because some neighborhood does not have any venues nearabout

### Now we will make model to cluster the Neighborhoods

In [24]:
from sklearn.cluster import KMeans

In [25]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [26]:
kmeans.labels_[0:50]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])

We will add clusters to the top 5 venue list

In [27]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [28]:
neighborhoods_venues_sorted.head()

Cluster Labels                                    Neighbourhood  \
0               0                                        Agincourt   
1               0                           Alderwood, Long Branch   
2               0  Bathurst Manor, Wilson Heights, Downsview North   
3               0                                  Bayview Village   
4               0                Bedford Park, Lawrence Manor East   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant                Lounge          Skating Rink   
1                Pizza Place           Coffee Shop                   Pub   
2                Coffee Shop                  Bank         Shopping Mall   
3                       Café    Chinese Restaurant   Japanese Restaurant   
4             Sandwich Place    Italian Restaurant           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue  
0        Breakfast Spot        Clothing Store  
1        Sandwich Place                  Pool  
2        Ice Cream Shop      Sushi Restaurant  
3                  Bank      Department Store  
4      Sushi Restaurant             Juice Bar

Now we will merge this table with the combined_data table and prepare for plotting

In [30]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
5              Islington Avenue, Humber Valley Village  43.667856 -79.532242   
6                                       Malvern, Rouge  43.806686 -79.194353   
7                                            Don Mills  43.745906 -79.352188   
8                      Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
9                             Garden District, Ryerson  43.657162 -79.378937   
10                                           Glencairn  43.709577 -79.445073   
11   West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724   
12              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
13                                           Don Mills  43.725900 -79.340923   
14                                    Woodbine Heights  43.695344 -79.318389   
15                                      St. James Town  43.651494 -79.375418   
16                                  Humewood-Cedarvale  43.693781 -79.428191   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201   
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
19                                         The Beaches  43.676357 -79.293031   
20                           

In [37]:
# we will remove the NaN values to prevent data skew
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Now we will plot the clusters on the map

In [36]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [ ]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### Now lets verify each of the clusters

**Cluster 1:**

In [38]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1          North York             0.0      Portuguese Restaurant   
2    Downtown Toronto             0.0                Coffee Shop   
3          North York             0.0     Furniture / Home Store   
4    Downtown Toronto             0.0                Coffee Shop   
6         Scarborough             0.0       Fast Food Restaurant   
7          North York             0.0                        Gym   
8           East York             0.0                Pizza Place   
9    Downtown Toronto             0.0                       Café   
10         North York             0.0                 Playground   
11          Etobicoke             0.0                     Bakery   
13         North York             0.0                        Gym   
14          East York             0.0               Skating Rink   
15   Downtown Toronto             0.0                  Gastropub   
16               York             0.0               Tennis Court   
17          Etobicoke             0.0                  Pet Store   
18        Scarborough             0.0        Rental Car Location   
19       East Toronto             0.0          Health Food Store   
20   Downtown Toronto             0.0                Coffee Shop   
22        Scarborough             0.0                Coffee Shop   
23          East York             0.0                Coffee Shop   
24   Downtown Toronto             0.0                Coffee Shop   
25   Downtown Toronto             0.0              Grocery Store   
26        Scarborough             0.0                Gas Station   
27         North York             0.0                Golf Course   
28         North York             0.0                Coffee Shop   
29          East York             0.0             Sandwich Place   
30   Downtown Toronto             0.0                Coffee Shop   
31       West Toronto             0.0                     Bakery   
32        Scarborough             0.0                 Playground   
33         North York             0.0                Coffee Shop   
..                ...             ...                        ...   
69       West Toronto             0.0                     Bakery   
70          Etobicoke             0.0                Pizza Place   
71        Scarborough             0.0                Auto Garage   
72         North York             0.0                Coffee Shop   
73    Central Toronto             0.0             Clothing Store   
74    Central Toronto             0.0             Sandwich Place   
75       West Toronto             0.0             Breakfast Spot   
76        Mississauga             0.0                Coffee Shop   
77          Etobicoke             0.0             Sandwich Place   
78        Scarborough             0.0  Latin American Restaurant   
79    Central Toronto             0.0               Dessert Shop   
80   Downtown Toronto             0.0                       Café   
81       West Toronto             0.0                       Café   
82        Scarborough             0.0       Fast Food Restaurant   
84   Downtown Toronto             0.0                       Café   
85        Scarborough             0.0                       Park   
86    Central Toronto             0.0                Coffee Shop   
87   Downtown Toronto             0.0             Airport Lounge   
88          Etobicoke             0.0               Liquor Store   
89          Etobicoke             0.0              Grocery Store   
90        Scarborough             0.0       Fast Food Restaurant   
92   Downtown Toronto             0.0         Seafood Restaurant   
93          Etobicoke             0.0                Pizza Place   
94          Etobicoke             0.0        Rental Car Location   
96   Downtown Toronto             0.0                       Café   
97   Downtown Toronto             0.0                       Café   
98          Etobicoke             0.0                      River   
99

**Cluster 2:**

In [39]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  Scarborough             1.0                   Bar           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
12    Falafel Restaurant           Escape Room     Electronics Store

**Cluster 3:**

In [40]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             2.0                  Park   
21              York             2.0                  Park   
35         East York             2.0                  Park   
52        North York             2.0                  Park   
64              York             2.0                  Park   
66        North York             2.0                  Park   
91  Downtown Toronto             2.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio         Deli / Bodega   
21         Women's Store                  Pool           Curling Ice   
35     Convenience Store          Intersection           Yoga Studio   
52           Yoga Studio          Dance Studio     Electronics Store   
64     Convenience Store           Yoga Studio         Deli / Bodega   
66     Convenience Store           Yoga Studio         Deli / Bodega   
91            Playground                 Trail          Dance Studio   

          5th Most Common Venue  
0             Electronics Store  
21  Eastern European Restaurant  
35                Deli / Bodega  
52  Eastern European Restaurant  
64            Electronics Store  
66            Electronics Store  
91            Electronics Store

**Cluster 4:**

In [41]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
57     Falafel Restaurant           Escape Room     Electronics Store  
101    Falafel Restaurant           Escape Room     Electronics Store

**Cluster 5:**

In [42]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto             4.0            Restaurant   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
83          Dance Studio     Electronics Store  Eastern European Restaurant   

   5th Most Common Venue  
83             Drugstore

**We have successfully clustered the Toronto Neighborhood based on the venue catagories**

### End of Assignment